File: devbak/status.py
from bkenv import *
from dhash import *
from tstamp import *
from status import *


def dirty1(T=None):
    yielded = {}
    for Npl, Op in opdep.items():
        for di, si in Npl:
            (N2, N1) = ts2(di, si)
            try:
                if bctck(N2, N1) and \
                    (T is None or T == di) and \
                        (Op not in yielded):
                    yield Op
                    yielded[Op] = True
            except:
                print(di, si, N2, N1)


def stsupdate(Si, Dh):
    print(Si, end=' ')
    N1 = srcts[Si]
    rtset(N1)
    dhset(Si, Dh)


def onestatus(Si):
    (Dh, changed) = dhck(Si)
    if changed:
        stsupdate(Si, Dh)


def statuses():
    SDl = []
    for Si in srcs:
        (Dh, changed) = dhck(Si)
        if changed:
            SDl.append((Si, Dh))
    return SDl


def statuses():
    SDl = []
    for Si in srcs:
        (Dh, changed) = dhck(Si)
        if changed:
            SDl.append((Si, Dh))
    return SDl


def updatets(N):
    print('Status', N)
    Sl = statuses()
    if len(Sl):
        print("changed: ", end='')
        for (Si, Dh) in Sl:
            stsupdate(Si, Dh)
        print()


if __name__ == '__main__':
    updatets(1)
    for i in dirty1():
        print(i)

File: prolog/old/status.pl
%#!/data/data/com.termux/files/usr/bin/env swipl

:- writeln("-- status.pl --").
:- ensure_loaded(["gb_env.pl", "funcs.pl", "dhash.pl"]).

dirty1(Op, Di, Si, Dd, Sd, Opl) :-
    opdep(Op, Di, Si, Dd, Sd, Opl),
    member(X, Di),
    member(Y, Si),
    ts2(X, Y, N2, N1),
    bctck(N2, N1).

stsupdate(dh(Si, Dh)) :-
    write(" "),
    write(Si),
    srcts(Si, N1),
    rtset(N1),
    !,
    dhset(Si, Dh).

mdp([]) :-
    nl,
    !.
mdp([H|T]) :-
    stsupdate(H),
    mdp(T).

statuses(SDl) :-
    findall(dh(Si, _), src(Si), Pl),
    include(dhck, Pl, SDl).

updatets(N) :-
    (   statuses(Sl),
        write("Status "),
        write(N),
        Sl\=[]
    ->  write(" changed:"),
        mdp(Sl)
    ;   nl
    ).

/*
depchk(T, D) :- dep(T, S1), depchk(S1, D1), D is D1 + 1, !.
depchk(_,0) :- !.

ddepth(D2) :- 
  findall(D1,(dep(T,_), depchk(T,D1)),L),
  print(L),
  max_list(L,D2).

maxd(Op, D2) :- 
  findall(D1, (opdep(Op, T, _, _, _, _), depchk(T,D1)), L),
  max_list(L,D2).

dcmp(<, _, _, D1, D2) :- D1 < D2, write(<).
dcmp(<, E1, E2, D1, D2) :- D1 == D2, E1 @< E2, write(<).
dcmp(=, E1, E2, D1, D2) :- D1 == D2, E1 == E2, write(=).
dcmp(>, E1, E2, D1, D2) :- D1 == D2, E1 @> E2, write(>).
dcmp(>, _, _, D1, D2) :- D1 > D2, write(>).

opcmp(Delta, E1, E2) :-  maxd(E1,D1), maxd(E2,D2), dcmp(Delta, E1, E2, D1, D2).

oplst(L2) :- 
  findall(Op,dirty1(Op, _, _, _, _, _),L1),
  write(L1), nl,
  predsort(opcmp,L1,L2),
  write(L2), nl,!.
*/
